pkg/instance/instance.go

```diff
@@ -447,18 +455,40 @@ func (s *Server) watchSPDKReplica(ctx context.Context, req *emptypb.Empty, srv r
 		return errors.Wrap(err, "failed to create SPDK replica watch notifier")
 	}
 
+	failureCount := 0
 	for {
+		if failureCount >= maxMonitorRetryCount {
+			logrus.Errorf("Continuously receiving errors for %v times, stopping watching SPDK replicas", maxMonitorRetryCount)
+			return fmt.Errorf("continuously receiving errors for %v times, stopping watching SPDK replicas", maxMonitorRetryCount)
+		}
+
 		select {
 		case <-ctx.Done():
-			logrus.Info("Stopped watching SPDK replica")
-			return nil
+			logrus.Info("Stopped watching SPDK replicas")
+			return ctx.Err()
 		default:
-			_, err := notifier.Recv()
-			if err != nil {
-				logrus.WithError(err).Error("Failed to receive next item in SPDK replica watch")
-			} else {
-				if err := srv.Send(&empty.Empty{}); err != nil {
-					return errors.Wrap(err, "failed to send instance response")
+			// Use a channel to receive the result of notifier.Recv()
+			notifyRecvChan := make(chan error, 1)
+
+			go func() {
+				_, err := notifier.Recv()
+				notifyRecvChan <- err
+			}()
+
+			select {
+			case <-ctx.Done():
+				// If context is done before notifier.Recv() completes, return immediately
+				logrus.Info("Stopped watching SPDK replicas")
+				return ctx.Err()
+			case err := <-notifyRecvChan:
+				if err != nil {
+					logrus.WithError(err).Error("Failed to receive next item in SPDK replica watch")
```

Comment:
suggest use `Warn` here and below since it is still acceptable before `failureCount` is larger than  `maxMonitorRetryCount`
We already `Error` the message above when `failureCount > maxMonitorRetryCount`

BTW, does the failure occur randomly, or will it continue to happen if it occurs once ?

Reply:
> BTW, does the failure occur randomly, or will it continue to happen if it occurs once ?

Both are possible.